Add support for torch.tensor_split to accept a tensor for `indices` argument

edqwerty10 opened the pull request:
Summary:
Trying to solve PR request https://github.com/pytorch/pytorch/issues/47479.
This diff tries to overload method `torch.tensor_split` to also accept a tensor for argument `split_size_or_sections` which currently accepts a python list or int. The motivation is to avoid converting a tensor to a list so that when tracing a model/module the tensor operations can be recorded.

Implementation is following the diff that originally added the `tensor_split` method D24166164 (https://github.com/pytorch/pytorch/commit/ef4817fe5a16ba9969562911c5363736a1003bb0).

Test Plan:
```
buck test caffe2/test:others -- test_tensor_split
buck test caffe2/test:torch -- test_tensor_split_indices
 buck test caffe2/test:torch -- test_tensor_split_errors
```

Differential Revision: D25440885

ngimel commented:
Test errors are real. Do you need help with those, @edqwerty10, or are you still working on the PR?

ngimel commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -545,6 +545,21 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
   return splits;
 }
 
+std::vector<Tensor> tensor_split(const Tensor& self, const Tensor& tensor_indices_or_sections, int64_t dim) {
```
please check that `indices_or_sections` is on cpu, it won't work otherwise (first branch will cause host-device sync, second branch will segfault when you are creating std::vector from tensor data)

edqwerty10 replied:
Hi @ngimel , how would the code look like to check `indices_or_sections` is on cpu? Is this a TORCH_CHECK?
Once `indices_or_sections` is confirmed to be on cpu does the body of the code change?

I think we want to support this for gpu as well, how would the code change to support this?
cc. @xw285cornell

ngimel replied:
Yes, it should be a TORCH_CHECK, something like
```
TORCH_CHECK(tensor_indices_or_sections.device()==kCPU, "expected tensor_indices_or_sections to be on cpu")
```
After the check the body of the code does not need to change. Allowing `indices_or_sections` to be on gpu would require you to copy it to cpu here, which is a cross-device synchronization. Pytorch tries not to do implicit cross-device synchronization, if you need it, you can do it in the client code by calling `indices.cpu()`

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -545,6 +545,23 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
+std::vector<Tensor> tensor_split(const Tensor& self, const Tensor& tensor_indices_or_sections, int64_t dim) {
+  TORCH_CHECK(self.dim() > 0, "expected at least a 1-dimensional tensor");
+  TORCH_CHECK(tensor_indices_or_sections.device() == kCPU,
+    "expected tensor_indices_or_sections to be on cpu")
```
"expected tensor_indices_or_sections to be on cpu" -> "tensor_split() expected tensor_indices_or_sections to be on cpu, but it's on DEVICE_TYPE"

mruberry replied:
(where DEVICE_TYPE is populated with the current device of `tensor_indices_or_sections`)

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -545,6 +545,23 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
+    return self.tensor_split(sections, dim);
+  } else {
+    TORCH_CHECK(tensor_indices_or_sections.dim() == 1 &&
+      tensor_indices_or_sections.scalar_type() == at::kLong,
```
This check for kLong is good but should happen immediately before or after the check that `tensor_indices_or_sections` is on the CPU. Otherwise a scalar float tensor could be passed.

mruberry replied:
The check for dimensions can also happen along with the other checks. Check that the dimension is <= 1, and if it's not say it's expected to be zero or one dimensions and print how many dimensions it actually is, just like the comment for the device warning above.

mruberry commented on test/test_view_ops.py:
```diff
@@ -1206,9 +1216,18 @@ def test_tensor_split_errors(self, device):
             with self.assertRaises(numpy_err, msg=msg):
                 np.array_split(a.cpu().numpy(), sections_or_indices, dim)
 
+        # test tensor indices being a 1-dimension tensor of dtype long
```
This section of the tests will need to be updated to reflect the error checking changes mentioned above

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1352,37 +1352,50 @@ def merge_dicts(*dicts):
 Args:
     input (Tensor): the tensor to split
     indices_or_sections (int or (list(int))):
```
`indices_or_sections (int or (list(int))):` -> `indices_or_sections (Tensor, int or list or tuple of ints):`

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1352,37 +1352,50 @@ def merge_dicts(*dicts):
-        dimension :attr:`dim` at each of the indices in the list. For instance,
-        :code:`indices_or_sections=[2, 3]` and :code:`dim=0` would result in the tensors
-        :code:`input[:2]`, :code:`input[2:3]`, and :code:`input[3:]`.
+        If :attr:`indices_or_sections` is an integer ``n`` or scalar tensor with value
```
"... or scalar tensor ..." -> "... or a zero dimensional long tensor ..."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1352,37 +1352,50 @@ def merge_dicts(*dicts):
+        sections will have size :code:`int(input.size(dim) / n) + 1`, and the rest will
+        have size :code:`int(input.size(dim) / n)`.
+
+        If :attr:`indices_or_sections` is a list of ints or a one-dimensional tensor of ints,
```
"... or a one-dimensional tensor of ints ..." -> "... or a one-dimensional long tensor ..."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1352,37 +1352,50 @@ def merge_dicts(*dicts):
+
+        If :attr:`indices_or_sections` is a list of ints or a one-dimensional tensor of ints,
+        :attr:`input` is split along dimension :attr:`dim` at each of the indices
+        in the list. For instance, :code:`indices_or_sections=[2, 3]` and :code:`dim=0`
```
"in the list" -> "in the list, tuple or tensor."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1352,37 +1352,50 @@ def merge_dicts(*dicts):
     >>> x = torch.arange(8)
     >>> torch.tensor_split(x, 3)
     (tensor([0, 1, 2]), tensor([3, 4, 5]), tensor([6, 7]))
+    >>> torch.tensor_split(x, torch.tensor(3))
```
These new examples are thorough but I think it's easy to infer what the behavior with passing a tensor will be. We probably don't need to change the examples.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -290,6 +290,13 @@ def np_sinc_with_fp16_as_fp32(x):
+    return (SampleInput(make_tensor((S, S, S), device, dtype,
+                                    low=None, high=None,
+                                    requires_grad=requires_grad),
+                        args=(make_tensor((S,), device, torch.int64, low=None, high=None, requires_grad=False),),
```
I think this is the cause of the test failures. The splits must be on the CPU, but here they're being put on the passed device. Changing device to be "cpu" should work.

mruberry replied:
There are a few other things to change with this SampleInput:

- add a second sample input where the indices_or_sections is a scalar tensor
- add a third sample input where the dim argument is used
- just create the indices_or_sections tensors directly instead of using `make_tensor`, that is, just call `torch.tensor(...` so the splits make sense given the input tensor. If the splits are randomly generated, for example, they might look very strange, like `(-5, 2, 1, ...)`

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1352,37 +1352,50 @@ def merge_dicts(*dicts):
+        :attr:`input` is split along dimension :attr:`dim` at each of the indices
+        in the list. For instance, :code:`indices_or_sections=[2, 3]` and :code:`dim=0`
+        would result in the tensors :code:`input[:2]`, :code:`input[2:3]`, and :code:`input[3:]`.
 
```
Add a new paragraph here that's just one sentence explaining the requirement that if this is a tensor it must be a CPU long tensor. Something like:

"If indices_or_sections is a tensor, it must be a zero-dimensional or one-dimensional long tensor on the CPU."

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -787,6 +794,16 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+           sample_inputs_func=sample_inputs_tensor_split,
+           skips=(
+               # Test is failing and reason is unknown so far
+               SkipInfo('TestCommon', 'test_variant_consistency_jit', dtypes=[torch.bfloat16]),
```
#49356 should fix this, but it's ok to skip the test if we don't land it in time.

mruberry replied:
#49356 has been merged. Try running the tests without this skip: it may no longer be needed.

mruberry commented:
Hey @edqwerty10, this is looking pretty good. I made some comments/suggestions inline, mostly about handling sample inputs and error checking. This also needs to be rebased because of changes in common_methods_invocations.py.

mruberry commented:
The failures are real. 

- make_tensor now has to take an actual torch.device, which can be constructed like `torch.device('cpu')`. My mistake on suggesting just "cpu" would do.
- looks like bfloat16 is supported on CUDA, do the dtypesIfCUDA needs to be updated to include bfloat16

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1252,19 +1252,22 @@ def merge_dicts(*dicts):
+        would result in the tensors :code:`input[:2]`, :code:`input[2:3]`, and :code:`input[3:]`.
+
+        If indices_or_sections is a tensor, it must be a zero-dimensional or one-dimensional
+        long tensor on the CPU."
```
Minus the quotation mark at the end of this sentence.

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -550,6 +550,26 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
+    int64_t sections = tensor_indices_or_sections.item<int64_t>();
+    return self.tensor_split(sections, dim);
+  } else {
+    TORCH_CHECK(split_dim == 1, "tensor_split expected indices to be a 1-dimensional tensor");
```
This check should still be moved up or refined, because tensor_split expects `tensor_indices_or_sections` to be either a zero dimensional or one dimensional tensor. There's no user-facing concept of "indices" that the message can reference.

If the check is moved outside this if statement, for example, it would very that tensor_indices_or_sections has zero or 1 dimensions.

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -550,6 +550,26 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
 }
 
+std::vector<Tensor> tensor_split(const Tensor& self, const Tensor& tensor_indices_or_sections, int64_t dim) {
+  TORCH_CHECK(self.dim() > 0, "expected at least a 1-dimensional tensor");
```
Same type of formatting here where it starts with "tensor_split":

"tensor_split expected input to have at least one dimension, but got ..."

edqwerty10 replied:
I kept it as is because the other two `tensor_split` signatures are raising the same exception message. If we change it, shall I change the other two signatures' messages as well?

mruberry replied:
OK, I appreciate that for consistency we can leave this as-is and not bother changing it.

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -515,7 +515,7 @@ std::vector<Tensor> chunk(const Tensor& self, int64_t chunks, int64_t dim) {
 
 std::vector<Tensor> tensor_split(const Tensor& self, int64_t sections, int64_t dim) {
-  TORCH_CHECK(self.dim() > 0, "expected at least a 1-dimensional tensor");
+  TORCH_CHECK(self.dim() > 0, "tensor_split expected at least a 1-dimensional tensor, got dim ", self.dim());
```
"got" -> "but got"

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -531,7 +531,7 @@ std::vector<Tensor> tensor_split(const Tensor& self, int64_t sections, int64_t d
 
 std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_t dim) {
-  TORCH_CHECK(self.dim() > 0, "expected at least a 1-dimensional tensor");
+  TORCH_CHECK(self.dim() > 0, "tensor_split expected at least a 1-dimensional tensor, got dim ", self.dim());
```
"got" -> "but got"

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -545,6 +545,28 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
 }
 
+std::vector<Tensor> tensor_split(const Tensor& self, const Tensor& tensor_indices_or_sections, int64_t dim) {
+  TORCH_CHECK(self.dim() > 0, "tensor_split expected at least a 1-dimensional tensor, got dim ", self.dim());
```
"got -> but got"

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -545,6 +545,28 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
+    "tensor_split expected tensor_indices_or_sections to have dtype of long, got ", split_dtype);
+  auto split_dim = tensor_indices_or_sections.dim();
+  TORCH_CHECK(split_dim == 1 || split_dim == 0,
+    "tensor_split expected tensor_indices_or_sections to be a zero-dimensional or one-dimensional tensor, got dim ", split_dim);
```
"got -> but got"

mruberry commented on test/test_view_ops.py:
```diff
@@ -1181,23 +1185,29 @@ def test_tensor_split_indices(self, device, dtype):
             # input size, sections or indices, dim, error type, error message, numpy error type
             [(S,), 10, 1, IndexError, r'Dimension out of range', IndexError],
-            [(), 10, 0, RuntimeError, r'expected at least a 1-dimensional tensor', IndexError],
+            [(), 10, 0, RuntimeError, r'tensor_split expected at least a 1-dimensional tensor, got dim 0', IndexError],
```
"got" language will need to be updated here, too

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -545,6 +545,28 @@ std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_
+    "tensor_split expected tensor_indices_or_sections to be on cpu, but its on ", split_device);
+  auto split_dtype = tensor_indices_or_sections.scalar_type();
+  TORCH_CHECK(split_dtype == at::kLong,
+    "tensor_split expected tensor_indices_or_sections to have dtype of long, got ", split_dtype);
```
"got -> "but got"

The "but" is typical (like in the error message above) to emphasize the surprise at not getting what's expected.

mruberry commented on test/test_view_ops.py:
```diff
@@ -1206,9 +1216,21 @@ def test_tensor_split_errors(self, device):
 
+        # addtional tests for tensor_split with tensor_indices_or_sections
+        with self.assertRaisesRegex(RuntimeError,
+                                    r'tensor_split expected tensor_indices_or_sections to have dtype of long, got Float'):
```
These regexes will need updating per the above changes, too

mruberry reviewed (approved):
Nice work, @edqwerty10! I made a few final comments about some error message formatting that should be straightforward to implement.

In addition to those changes, let @ngimel take a look before merging this, too.

Commit:
Add support for torch.tensor_split to accept a tensor for `indices` argument (#49169)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/49169

Trying to solve PR request https://github.com/pytorch/pytorch/issues/47479.
This diff tries to overload method `torch.tensor_split` to also accept a tensor for argument `split_size_or_sections` which currently accepts a python list or int. The motivation is to avoid converting a tensor to a list so that when tracing a model/module the tensor operations can be recorded.

Implementation is following the diff that originally added the `tensor_split` method D24166164 (https://github.com/pytorch/pytorch/commit/ef4817fe5a16ba9969562911c5363736a1003bb0).

Test Plan:
```
buck test caffe2/test:torch -- tensor_split
```
https://www.internalfb.com/intern/testinfra/testconsole/testrun/5910974550563805/

```
buck test caffe2/test:others -- tensor_split
```
https://www.internalfb.com/intern/testinfra/testconsole/testrun/1688849905082678/

Reviewed By: mruberry

Differential Revision: D25440885

fbshipit-source-id: ca5d134cfb91fa0efc3dec5257dbc97532eb2d74

edqwerty10 commented:
Thanks a lot for the help @mruberry  and @ngimel! :) I made the amends and I'm looking to land the PR/diff after the tests complete (amends will reflect on github soon). Do I have permission to do this? I see `Merging is blocked` which makes me think I do not. I can still hit the land button from Phabricator but looks like we need to merge from github side too, I think.

mruberry commented:
> Thanks a lot for the help @mruberry and @ngimel! :) I made the amends and I'm looking to land the PR/diff after the tests complete (amends will reflect on github soon). Do I have permission to do this? I see `Merging is blocked` which makes me think I do not. I can still hit the land button from Phabricator but looks like we need to merge from github side too, I think.

You only need to "Ship It" from Phabricator once the tests finish and an automated system will mirror the changes to Github.
